pygfx/helpers/_gizmo.py

```diff
@@ -0,0 +1,747 @@
+"""
+A transform gizmo to manipulate world objects.
+"""
+
+import numpy as np
+
+from ..objects import Line, Mesh
+from ..geometries import Geometry, sphere_geometry, cone_geometry, box_geometry
+from ..materials import MeshBasicMaterial, LineMaterial
+from ..objects import WorldObject
+from ..linalg import Vector3, Quaternion
+
+
+# Colors in hsluv space - https://www.hsluv.org/
+# With H: 0/120/240, S: 100, L: 50
+RED = "#ea0064"
+GREEN = "#3f8700"
+BLUE = "#007eb7"
+
+
+class TransformGizmo(WorldObject):
+    """
+    A gizmo object that can be used to manipulate (i.e. transform) a world object.
+
+    Aguments:
+        object (WorldObject): the object to transform with the gizmo.
+        screen_size (float): the approximate size of the widget in logical pixels. Default 100.
+
+    To control the Gizmo:
+    * Click the center sphere to toggle between object-space, world-space and screen-space.
+    * Grab the center sphere for uniform scaling.
+    * Grab the cubes for one-directional scaling (only in object-space).
+    * Grab the arrows to translate in one direction.
+    * Grab the planes to translate in two directions.
+    * Grab the spheres to rotate.
+    """
+
+    def __init__(self, object=None, screen_size=100):
+        super().__init__()
+
+        # We store these as soon as we get a call in handle_event
+        self._renderer = None
+        self._canvas = None
+        self._camera = None
+
+        # A dict that stores the state at the start of a drag. Or None.
+        self._ref = None
+
+        # The (approximate) size of the gizmo on screen
+        self._screen_size = float(screen_size)
+
+        # Init
+        self.set_object(object)
+        self._create_elements()
+        self.toggle_mode("object")  # object, world, screen
+        self._highlight()
+
+    def set_object(self, object):
+        """Set the WorldObject to control with the gizmo."""
+        if object is None:
+            self._object_to_control = None
+        else:
+            assert isinstance(object, WorldObject)
+            self._object_to_control = object
+
+    def toggle_mode(self, mode=None):
+        """Toggle between modes. If the mode is omitted, will move to the next mode.
+        Must be one of "object", "world", or "screen".
+        """
+        modes = "object", "world", "screen"
+        if not mode:
+            mode = {"object": "world", "world": "screen"}.get(self._mode, "object")
```

Comment:
this is the first instance I've seen where `itertools.cycle` would actually be useful :) anyway, it's fine!